Add high-level Conf and Analysis classes

Bultako opened the pull request:
This PR is the first of a series related with [PIG 12: High-level interface](https://github.com/gammapy/gammapy/pull/2219)

It introduces the Analysis class as a session class. At this moment it only does handling of the setting and validation of the analysis session configuration parameters. The main points to highlight are:

* [jsonschema](https://github.com/Julian/jsonschema) dependency for validation of configuration parameters against a defined schema
* validation of astropy quantities and units for the values of the configuration parameters
*  user setting, modification and removal of configuration parameters using the interface and/or providing a user defined configuration file
* display of the actual configuration parameters of the session 
* reset method to recover the original configuration state

cdeil commented:
@Bultako - Thanks!

This is very hard to review. I see a complex Singleton class, but it's not clear why it's needed or why we would want that, and also other code like the `reset` method that isn't used or tested and not clear to me if we want that. Can you please delete that, or show a useful docs or test example?

Like I said in the call this morning, I would suggest to always do use-case and test-driven development, and as a first step to try to drive this via the new code:
https://github.com/gammapy/gammapy/blob/master/gammapy/scripts/spectrum_pipe.py
and to migrate the existing test case for that to the new scheme and then to delete that code.
At the start what you get will be a mix of using the config file and hard-coded methods that do stuff based on config file content, I think that's fine and even if we manage to get something better that would be the path forward.

@Bultako - Where do you want to go with this PR? Continue to a point where this drives an analysis (either the spectral analysis, or what `gammapy bin cube` does now to have a simpler example), or merge this soon, and continue in a new PR?

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -0,0 +1,131 @@
+        """Reset configuration parameters to initial state."""
+        self.__init__(self.configfile)
+
+    def list_config(self):
```
`list_config` sounds like a Python list. But this is a dict and hierarchical data structure, no?
Suggest to call the data member `config` and have it be a dict or dict-like object that represents the config.

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -0,0 +1,131 @@
+        The name of a user defined configuration file.
+    """
+
+    __slots__ = ["configfile", "_params"]
```
Why use slots? Suggest to remove slots and use normal Python attributes.
(we use slots in other places, probably not a great idea / coding pattern)

cdeil commented:
I didn't look at the design of Fermipy or ctapipe or other packages for config-handling.
Would it make sense to have an `AnalysisConfig` class that mostly does the schema validation, merging in defaults with user-defined settings etc.?
And then the Analysis class would just make one of those, and it's job would be to orchestrate execution based on a config?
I think this separation of concerns into two classes might be simpler than a single class?

Bultako commented:
@cdeil
Thanks very much for the review.
I think I've addressed most of your comments and suggestions.

I've greatly simplified the `Analysis` class.

I have completely switched to an optic of seeing users as *consenting adults*. This has reduced much of the complex code, so it also allow us to not consider (for the moment) having another class for configuration handling. As it is the case with Fermipy, we can instantiate the class with a config file or/and with config parameters passed at that moment. By default the `Analysis` class will consider a small set of config params declared in the `default.yaml` file, the user can also decide not to consider the default file, but his own file, also params passed as **kwargs (that will be rewrite those present in the config file) or neither file not params, and in that case she will be filling and modifying the config params of an *empty config* class during the analysis session. The validation of configuration params has to be done explicitly by the user at any moment with `analysis.validate_config()`, the implicit validation when setting/modifying params has been removed.

Re: `Analysis` class as a singleton
This may be an interesting point of discussion. I see `Analysis` also as a session class, logging actions in a single history, hence I wanted to be sure that no other `Analysis` class was opened in parallel to not miss any info of the history.  But it is also true that we could provide the possibility to use (maybe in scripts) several different Analysis classes in parallel. For now I've removed the singleton constraint, we could face this in the future if need arrives.

Since we do have now a smaller amount of code to review, I would prefer to merge this PR soon before starting a bigger one addressing a real use case, or at least agree on the code we have now.

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -0,0 +1,76 @@
+SCHEMA_FILE = CONFIG_PATH / "schema.yaml"
+
+
+class Analysis():
```
Remove redundant parentheses

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -0,0 +1,76 @@
+        schema = read_yaml(SCHEMA_FILE)
+        try:
+            jsonschema.validate(self.config, schema, _gp_validator)
+        except jsonschema.exceptions.ValidationError as ex:
```
Please remove try-except, or add a test for the except case. From just reading it, I have no idea if `ex.schema_path[-2]` will work or not, this kind of code that was only tested manually once is often where bugs hide.

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -0,0 +1,76 @@
+
+    def _set_logging(self):
+        """Set logging parameters for API."""
+        if "global" in self.config and "logging" in self.config["global"]:
```
We should get rid of the `if` here. If we start to write code like this everywhere we get a config value, we're in hell. I'll make a suggestion how to handle this below.

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -0,0 +1,76 @@
+            log.info("Setting logging parameters ({}).".format(self.config["global"]["logging"]["level"]))
+
+
+def _astropy_quantity(_, instance):
```
This `_astropy_quantity` is untested. I would suggest to delete it from this PR, and to introduce quantity parsing with tests in a separate PR. But if you want to add it here, OK, please add a test to show that & how it works.

Bultako replied:
Ok.
I've added a test to validate parameter settings with astropy quantities as values.
I think validating astropy quantities with jsonschema will be needed (or at least helpful) in next PRs.

cdeil replied:
ok

cdeil commented:
@Bultako - This is another project worth looking at IMO: https://github.com/mahmoud/glom
You could use it if you want, or just as an example.
Probably if we have defaults, we don't need any of the fallback logic, we can just write:
```
config["global"]["logging"]["level"]
```
and it's guaranteed to be there, because it's in the default config and we don't support users messing with the config, if they do, it can just fail with KeyError.

Having this I guess would be nice to have, but also causes problems because it's limited to string keys, so not really needed for now?
```
config.global.logging.level
config["global.logging.level"]
```

I think if you want to get good validation of config for setting e.g. for
```
config.global.logging.level = "spam"
```
at runtime isn't easy to achieve (but also not very hard), we'd need to have a `ConfigItem` proxy class or somehow trigger validation on setting. I think this is something we can develop later, for now we just have to try and get the config getting API right.

I also see that you use `**kwargs` in ``Analysis.__init__` to set config. I suggest to not use that coding pattern, but to take an explicit `config` dict argument instead. The problem with `**kwargs` is that then often what happens is that there's a mix of config things, and other arguments, and in the method complex logic to sort out the two accumulates. It's just as easy and cleaner to take this:
```
config = {...}
analysis = Analysis(config=config)
```

cdeil commented:
@Bultako - Merge conflict in `setup.py`. See https://github.com/gammapy/gammapy/pull/2279#issuecomment-525180658

Could you please rebase this branch?

I tried locally, and found that `git rebase --continue` didn't work for some reason and I had to use `git rebase --skip` instead. See [console log](https://gist.github.com/cdeil/c7cd728ec822f0ea8269dc5b02b04c12) and [result](https://github.com/gammapy/gammapy/compare/master...cdeil:HLI?expand=1).

Bultako commented:
@cdeil 
Thanks for all those suggestions.
 I summarize below the main actions of what I've done:

- Added `setup_package.py` to install the schema YAML file
- Added a `Config` separated class dealing with parameters/settings handling.
- Removed `default.yaml` file and use **only one** `schema.yaml` file to validate the settings, and also to declare **default values**. This has been done extending the jsonschema validator to use our own, that also implements validation of astropy quantities (I've added a test for this)
- Stored settings in three different internal properties of the `Config` class (default settings, settings from file, user defined settings) at the moment of instantiation of `Analysis`    

```
    Here are different examples on how to create an `Analysis` session class:

    >>> from gammapy.scripts import Analysis
    >>> cfg = {"general": {"out_folder": "myfolder"}}
    >>> analysis = Analysis(configfile="myfile.yaml", config=cfg)
    >>> analysis = Analysis(config=cfg)
    >>> analysis = Analysis()
``` 

I use a small `_update_settings` function of my own to **not overwrite all** the default settings by a user supplied dictionary of settings, but **only those specified by the user** in the `config` param or in his/her own configuration file. Yes, I think we might have a look at more fancy solutions for parameter handling, but I prefer to go ahead with a simple solution now and come back to this later when working with complex/real uses cases rise this need.

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -0,0 +1,134 @@
+
+
+class Analysis:
+    """The Analysis class drives an analysis working session using the high-level interface API.
```
Always put a short one-line summary, then an empty line, and then extra text in the docstring.
The summary line shows up in API docs summary tables.

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -0,0 +1,134 @@
+    >>> analysis = Analysis()
+    """
+
+    def __init__(self, configfile="", config=dict()):
```
Avoid mutable default arguments: see https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

I'm -1 to have The Analysis init have to deal with two ways to create a config.
Why not keep it simple and just take a `config`, and if someone has a config file, they write this:
```
config = Config.read("config.yaml")
analysis = Analysis(config)
```
Note that very soon Analysis will grow other things, because e.g. complex models and datasets and observations can't be created from config (maybe for now, maybe forever). So it'll really be nice if Analysis doesn't need any config-read related logic and arguments.

Overall config could have a default of an empty dict, but that should be in the Config class init, and it should use this pattern:
ttps://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

cdeil commented on gammapy/scripts/analysis.py:
```diff
@@ -0,0 +1,134 @@
+from astropy import units as u
+from pathlib import Path
+
+__all__ = ["Analysis"]
```
Suggest to add `Config` here and have it show up in the docs. It will grow functionality and users will interact with it, so I think it's appropriate to have it show up in the API docs.

cdeil commented on gammapy/scripts/setup_package.py:
```diff
@@ -0,0 +1,6 @@
+"""Add schema validation YAML file for High-level interface API"""
+
+
+def get_package_data():
```
Since today ( #2279 ) we maintain package data here:
https://github.com/gammapy/gammapy/blob/c5deffad757a69e26443986a8f0d94beda3d9af8/setup.cfg#L67

Could you please remove setup_package.py again and do the change in this PR?

Bultako commented:
@cdeil 
I think I've done all your suggestions, please check.
I write some examples to clarify what we have now for param settings handling :)

```
>>> analysis = Analysis()
>>> analysis.settings
{'general': {'logging': {'level': 'INFO'}, 'out_folder': '.'}}

>>> cfg = {"general": {"out_folder": "myfolder"}}
>>> analysis = Analysis(cfg)
>>> analysis.settings
{'general': {'logging': {'level': 'INFO'}, 'out_folder': 'myfolder'}}
```

Note how the default logging level info is still there in `analysis.settings["general"]`, only the `out_folder` setting has been modified. 

We can access and modify the settings from the analysis object, `analysis` has-a `configuration` and `settings`, so we can use `analysis.settings` and `analysis.configuration.validate()` to validate the config.

``` 
>>> analysis.settings["general"]["logging"]["level"] = "WARNING"
>>> analysis.configuration.validate()
```

But we need to proceed step-by-step to define nested properties
```
>>> analysis.settings["observations"]["datastore"]= "$GAMMAPY_DATA/cta/"
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-5-adc372c1634c> in <module>
----> 1 analysis.settings["observations"]["datastore"]= "$GAMMAPY_DATA/cta-1dc/"

KeyError: 'observations'
>>> analysis.settings["observations"] = {}
>>> analysis.settings["observations"]["datastore"] = "$GAMMAPY_DATA/cta-1dc/"
```

cdeil commented on gammapy/scripts/tests/test_analysis.py:
```diff
@@ -0,0 +1,28 @@
+from ..analysis import Analysis
```
Add standard license line like in all other Gammapy code files.

cdeil commented:
@Bultako - Thanks!

I've done some minor code formatting in f3ba360e3. (ran `make black` and `make isort` and renamed some variables and arguments to `config` instead of `cfg` or `configuration` to have less variation, with the goal for users to more easily remember the API.

I'm merging this now.

But FYI: I still think we should simplify and change the `Config`, to work like a collections.ChainMap with user settings before default settings on get, but still as separate objects so that defaults never get changed, but user settings can change. And we need tests for that. Once we have some usage and real-world experience with this class in the coming weeks, let's come back to this and try to simplify and improve the design.

Bultako commented:
@cdeil 
Thanks !

Re: user settings and default settings

I've tried doing the settings handling with `collections.ChainMap` to have the behavior that you expose. The problem is that we work with nested dictionaries, so with ChainMap the whole upper level dict is replaced by what it is defined in the user config dict, and the default values are lost.

i.e.
```
>>> analysis = Analysis()
>>> analysis.settings
{'general': {'logging': {'level': 'INFO'}, 'outdir': '.'}}

>>> config = {"general": {"outdir": "myfolder"}}
>>> analysis = Analysis(config)
>>> analysis.settings
{'general': {'outdir': 'myfolder'}}
```

That's why I wrote my own function, but I agree we will need to get back to this.
I also would like to have a better way to set the config with `analysis.settings`, instead of writing a list of brackets and avoiding defining empty values for deep undefined nested levels.
```
>>> analysis.settings["observations"]["datastore"]= "$GAMMAPY_DATA/cta/"
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-5-adc372c1634c> in <module>
----> 1 analysis.settings["observations"]["datastore"]= "$GAMMAPY_DATA/cta-1dc/"

KeyError: 'observations'
>>> analysis.settings["observations"] = {}
>>> analysis.settings["observations"]["datastore"] = "$GAMMAPY_DATA/cta-1dc/"
```

 I will revisit some of the links references you wrote above for inspiration.. But, yes parameter handling is still in my TODO list.
